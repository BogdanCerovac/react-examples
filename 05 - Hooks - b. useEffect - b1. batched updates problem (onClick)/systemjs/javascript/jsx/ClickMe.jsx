/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

/* eslint react/prop-types:"off" */

import React from 'react';

// SystemJS limitation (we should be able to do this directly in import)
const { useState } = React;

function ClickMe() {
  const [count, setCount] = useState(0);

  return (
    <a
      onClick={() => {
        setCount(count + 0.5);
        setCount(count + 0.5);
      }}
    >
      {`This link has been clicked ${count} times`}
    </a>
  );
}

export default ClickMe;
