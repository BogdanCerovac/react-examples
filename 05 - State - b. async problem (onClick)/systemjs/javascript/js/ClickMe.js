/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

/* eslint react/prop-types:"off" */

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // Updates are batched for performance reasons.
    // Therfor this won't work:
    this.setState({ count: this.state.count + 0.5 });
    this.setState({ count: this.state.count + 0.5 });
  }
  render() {
    return React.createElement('a', { onClick: this.onClick },
      'This link has been clicked ',
      this.state.count,
      ' times'
    );
  }
}
