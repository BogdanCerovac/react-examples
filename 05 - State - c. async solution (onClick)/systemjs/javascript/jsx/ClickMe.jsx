/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

/* eslint react/prop-types:"off" */

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // this.state may be updated asynchronously:
    this.setState(prevState => ({ count: prevState.count + 1 }));
  }
  render() {
    return (
      <a onClick={this.onClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}
