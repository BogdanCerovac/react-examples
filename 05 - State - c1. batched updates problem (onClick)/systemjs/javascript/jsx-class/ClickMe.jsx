/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // Updates are batched for performance reasons.
    // Therefore this won't work:
    this.setState({ count: this.state.count + 0.4 });
    this.setState({ count: this.state.count + 0.6 });
  }
  render() {
    return (
      <a onClick={this.onClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}
