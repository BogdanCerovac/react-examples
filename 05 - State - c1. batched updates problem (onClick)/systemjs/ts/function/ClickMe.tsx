/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

function ClickMe(): JSX.Element {
  const [count, setCount]: [
    number,
    React.Dispatch<React.SetStateAction<number>>
  ] = React.useState(0);

  const handleClick = () => {
    // Updates are batched for performance reasons.
    // Therefore this won't work:
    setCount(count + 0.4);
    setCount(count + 0.6);
  };

  return (
    <a onClick={handleClick}>{`This link has been clicked ${count} times`}</a>
  );
}

export default ClickMe;
