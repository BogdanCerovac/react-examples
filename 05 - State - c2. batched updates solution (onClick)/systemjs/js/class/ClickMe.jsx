/*! Mozilla Public License Version 2.0 !*/
/*! Copyright Â© 2018 Rick Beerendonk   !*/

import React from 'react';

export default class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    // Updates are batched for performance reasons.
    // Therefore this should be used:
    this.setState(state => ({ count: state.count + 0.4 }));
    this.setState(state => ({ count: state.count + 0.6 }));
  }
  render() {
    return (
      <a onClick={this.handleClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}
