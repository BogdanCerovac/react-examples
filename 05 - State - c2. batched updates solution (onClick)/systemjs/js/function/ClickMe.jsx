/*! European Union Public License version 1.2 !*/
/*! Copyright Â© 2018 Rick Beerendonk          !*/

import React from 'react';

function ClickMe() {
  const [count, setCount] = React.useState(0);

  return (
    <a
      onClick={() => {
        setCount(c => c + 0.4);
        setCount(c => c + 0.6);
      }}
    >
      This link has been clicked {count} times
    </a>
  );
}

export default ClickMe;
