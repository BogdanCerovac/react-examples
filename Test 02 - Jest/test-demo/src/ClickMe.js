import React from 'react';

class ClickMe extends React.Component {
  constructor(props) {
    super(props);
    this.state = { count: 0 };

    // Bind all non-react methods to this.
    this.onClick = this.onClick.bind(this);
  }
  onClick() {
    // this.state may be updated asynchronously:
    this.setState(state => ({ count: state.count + 1 }));
  }
  render() {
    return (
      <a onClick={this.onClick}>
        {`This link has been clicked ${this.state.count} times`}
      </a>
    );
  }
}

export default ClickMe;
