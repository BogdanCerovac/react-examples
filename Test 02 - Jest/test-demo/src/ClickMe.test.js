import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ClickMe from './ClickMe';

/// Enzyme ///

it('renders 0 snapshot (enzyme)', () => {
  const wrapper = shallow(<ClickMe />);
  expect(wrapper).toMatchSnapshot();
});

it('renders 1 after click snapshot (enzyme)', () => {
  const wrapper = shallow(<ClickMe />);
  wrapper.find('a').simulate('click');
  expect(wrapper).toMatchSnapshot();
});

/// React Test Renderer ///

it('renders 0 snapshot (react-test-renderer)', () => {
  const component = renderer.create(<ClickMe />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders 1 after click snapshot (react-test-renderer)', () => {
  const component = renderer.create(<ClickMe />);
  let tree = component.toJSON();
  tree.props.onClick();
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

