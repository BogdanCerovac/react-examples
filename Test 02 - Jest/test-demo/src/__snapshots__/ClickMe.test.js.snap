exports[`test renders 0 snapshot (enzyme) 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <a
    onClick={[Function]}>
    This link has been clicked 0 times
</a>,
  "nodes": Array [
    <a
      onClick={[Function]}>
      This link has been clicked 0 times
</a>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ClickMe />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ClickMe {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onClick": [Function],
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "count": 0,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <a
          onClick={[Function]}>
          This link has been clicked 0 times
</a>,
        "_debugID": 2,
        "_renderedOutput": <a
          onClick={[Function]}>
          This link has been clicked 0 times
</a>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ClickMe />,
}
`;

exports[`test renders 0 snapshot (react-test-renderer) 1`] = `
<a
  onClick={[Function]}>
  This link has been clicked 0 times
</a>
`;

exports[`test renders 1 after click snapshot (enzyme) 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <a
    onClick={[Function]}>
    This link has been clicked 1 times
</a>,
  "nodes": Array [
    <a
      onClick={[Function]}>
      This link has been clicked 1 times
</a>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ClickMe />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ClickMe {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onClick": [Function],
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "count": 1,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <a
          onClick={[Function]}>
          This link has been clicked 1 times
</a>,
        "_debugID": 4,
        "_renderedOutput": <a
          onClick={[Function]}>
          This link has been clicked 1 times
</a>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ClickMe />,
}
`;

exports[`test renders 1 after click snapshot (react-test-renderer) 1`] = `
<a
  onClick={[Function]}>
  This link has been clicked 1 times
</a>
`;
